plugin/util/parse_bindings.go

```diff
@@ -69,68 +69,122 @@ func parseBindingObjList(topList *ast.ObjectList) (map[string]StringSet, error)
-			continue
-		}
+func parseResourceObject(item *ast.ObjectItem, bindings map[string]StringSet) error {
+	if len(item.Keys) != 2 || item.Keys[0] == nil || item.Keys[1] == nil {
+		return fmt.Errorf(`top-level items must have format "resource" "$resource_name"`)
+	}
 
-		resourceName := item.Keys[1].Token.Value().(string)
-		_, ok := bindings[resourceName]
-		if !ok {
-			bindings[resourceName] = make(StringSet)
-		}
+	k, err := parseStringFromObjectKey(item, item.Keys[0])
+	if err != nil {
+		return err
+	}
+	if k != "resource" {
+		return fmt.Errorf(`invalid item %q, expected "resource"`, k)
+	}
+
+	resourceName, err := parseStringFromObjectKey(item, item.Keys[1])
+	if err != nil {
+		return err
+	}
+
+	_, ok := bindings[resourceName]
+	if !ok {
+		bindings[resourceName] = make(StringSet)
+	}
+	boundRoles := bindings[resourceName]
+
+	resourceItemList := item.Val.(*ast.ObjectType).List
+	if resourceItemList == nil {
+		return fmt.Errorf("invalid empty roles list for item", item.Assign.Line)
+	}
 
-		resourceList := item.Val.(*ast.ObjectType).List
-		for _, rolesItem := range resourceList.Items {
-			key := rolesItem.Keys[0].Token.Text
-			switch key {
-			case "roles":
-				parseRoles(rolesItem, bindings[resourceName], merr)
-			default:
-				merr = multierror.Append(merr, fmt.Errorf("invalid key '%s' in resource '%s' (line %d)", key, resourceName, item.Assign.Line))
-				continue
-			}
+	var merr *multierror.Error
+	for _, rolesObj := range resourceItemList.Items {
+		err := parseRolesObject(rolesObj, boundRoles)
+		if err != nil {
+			merr = multierror.Append(merr, fmt.Errorf("role list (line %d): %v", rolesObj.Assign.Line, err))
 		}
 	}
-	err := merr.ErrorOrNil()
+	return merr.ErrorOrNil()
+}
+
+func parseRolesObject(rolesObj *ast.ObjectItem, parsedRoles StringSet) error {
+	if rolesObj == nil || len(rolesObj.Keys) != 1 || rolesObj.Keys[0] == nil {
+		return fmt.Errorf(`expected "roles" list, got nil object item`)
+	}
+	k, err := parseStringFromObjectKey(rolesObj, rolesObj.Keys[0])
 	if err != nil {
-		return nil, err
+		return err
 	}
-	return bindings, nil
+	if k != "roles" {
+		return fmt.Errorf(`invalid key %q in resource, expected "roles"`, k)
+	}
+
+	if rolesObj.Val == nil {
+		return fmt.Errorf(`expected "roles" list, got nil value`)
+	}
+	roleList, ok := rolesObj.Val.(*ast.ListType)
+	if !ok {
+		return fmt.Errorf("parsing error, expected list of roles for key %q")
+	}
+	var merr *multierror.Error
+	for _, singleRoleObj := range roleList.List {
+		role, err := parseRole(rolesObj, singleRoleObj)
+		if err != nil {
+			merr = multierror.Append(merr, err)
+		} else {
+			parsedRoles.Add(role)
+		}
+	}
+	return merr.ErrorOrNil()
 }
 
-func parseRoles(item *ast.ObjectItem, roleSet StringSet, merr *multierror.Error) {
-	lst, ok := item.Val.(*ast.ListType)
+func parseRole(parent *ast.ObjectItem, roleNode ast.Node) (string, error) {
+	roleLitType, ok := roleNode.(*ast.LiteralType)
```

Comment:
ast.Node is an interface type which means it can be nil, so it should be nil-checked.

Reply:
@emilymye did you see Jeff's comment?

Reply:
@jefferai @sethvargo sorry fixing now - was on bug duty for TF last week so catching up on previous work now